plugins/arrow-flight-rpc/src/main/java/org/opensearch/arrow/flight/bootstrap/FlightClientManager.java

```diff
@@ -0,0 +1,280 @@
+            Objects.requireNonNull(elg, "EventLoopGroup cannot be null"),
+            Objects.requireNonNull(grpcExecutor, "ExecutorService cannot be null")
+        );
+        this.client = Objects.requireNonNull(client, "Client cannot be null");
+        clusterService.addListener(this);
+    }
+
+    /**
+     * Returns a Flight client for a given node ID.
+     *
+     * @param nodeId The ID of the node for which to retrieve the Flight client
+     * @return An OpenSearchFlightClient instance for the specified node
+     */
+    public OSFlightClient getFlightClient(String nodeId) {
+        return flightClients.containsKey(nodeId) ? flightClients.get(nodeId).flightClient : null;
+    }
+
+    /**
+     * Returns the location of a Flight client for a given node ID.
+     *
+     * @param nodeId The ID of the node for which to retrieve the location
+     * @return The Location of the Flight client for the specified node
+     */
+    public Location getFlightClientLocation(String nodeId) {
+        return flightClients.containsKey(nodeId) ? flightClients.get(nodeId).location : null;
+    }
+
+    /**
+     * Builds a client for a given nodeId in asynchronous manner
+     * @param nodeId nodeId of the node to build client for
+     */
+    public void buildClientAsync(String nodeId) {
+        CompletableFuture<Location> locationFuture = new CompletableFuture<>();
+        locationFuture.thenAccept(location -> {
+            DiscoveryNode node = getNodeFromClusterState(nodeId);
+            buildClientAndAddToPool(location, node);
+        }).exceptionally(throwable -> {
+            logger.error("Failed to get Flight server location for node: {}{}", nodeId, throwable);
+            return null;
+        });
+        requestNodeLocationAsyncAndBuildClient(nodeId, locationFuture);
+    }
+
+    @VisibleForTesting
+    void updateFlightClients() {
+        Set<String> currentNodes = getCurrentClusterNodes();
+        flightClients.keySet().removeIf(nodeId -> !currentNodes.contains(nodeId));
+        for (DiscoveryNode node : Objects.requireNonNull(clientConfig.clusterService).state().nodes()) {
+            buildClientAsync(node.getId());
+        }
+    }
+
+    Map<String, ClientHolder> getClients() {
+        return flightClients;
+    }
+
+    private void buildClientAndAddToPool(Location location, DiscoveryNode node) {
+        if (!isValidNode(node)) {
+            logger.warn(
+                "Unable to build FlightClient for node [{}] with role [{}] on version [{}]",
+                node.getId(),
+                node.getRoles(),
+                node.getVersion()
+            );
+            return;
+        }
+        OSFlightClient flightClient = buildClient(location);
+        flightClients.put(node.getId(), new ClientHolder(location, flightClient));
+    }
+
+    private void requestNodeLocationAsyncAndBuildClient(String nodeId, CompletableFuture<Location> future) {
```

Comment:
I think this can just be `requestNodeLocationAsync` since the call to buildClient is happening in `buildClientAsync`